misc/random.c

```diff
@@ -0,0 +1,66 @@
+/*
+ * Implementation of non-cryptographic pseudo-random number
+ * generator algorithm known as xoshiro.
+ *
+ * This file is part of mpv.
+ *
+ * mpv is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * mpv is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdint.h>
+
+#include "random.h"
+
+static volatile uint64_t s[4];
```

Comment:
this is not thread-safe as is, adding `volatile` here is just placebo

Reply:
It's not supposed to make it thread-safe, it's supposed to prevent an OS from creating a local thread-cache.

Reply:
Also, if I don't mark this one as volatile, then I get mismatched qualifier issues with regard to the one inside the struct itself.

Reply:
Thread safety is still a possible concern if multiple threads call `mp_rand_next()` at the same time, right? Since you don't lock the PRNG state.

Reply:
Is there any better way to do it though than marking it as volatile? I can add a comment that says the volatile keyword doesn't guarantee thread-safety.

Reply:
The compiler and or OS cannot possibly "create a thread-cached version" of a global variable without violating the C standard (or common sense). `volatile` has no effect on processor caches either, in case you meant that.

Reply:
In that case I'll just remove the volatile keyword in the interest of simplicity.